src/vsr/replica.zig

```diff
@@ -3920,16 +3920,23 @@ pub fn ReplicaType(
             assert(self.commit_min <= self.op);
             maybe(self.commit_max <= self.op);
 
-            if (!self.valid_hash_chain(@src())) {
-                assert(!self.solo());
-                return .ready;
-            }
-
             // We may receive commit numbers for ops we do not yet have (`commit_max > self.op`):
             // Even a naive state sync may fail to correct for this.
             if (self.commit_min < self.commit_max and self.commit_min < self.op) {
                 const op = self.commit_min + 1;
-                const header = self.journal.header_with_op(op).?;
+                const header = self.journal.header_with_op(op) orelse return .ready;
+
+                // Assuming that the head op is correct, it is definitely safe to commit the next
+                // prepare if it is from the same view as the head --- the primary for that view
+                // made sure that the hash chain is valid. If it is from the different view, we
+                // additionally verify ourselves that the hash-chain is not broken
+                const safe_to_commit = self.valid_hash_chain(@src()) or (self.status == .normal and
```

Comment:
Further restrict this to backups?

Reply:
It is restricted to backups, we are under `assert(!(self.status == .normal and self.primary()));`

Reply:
Yes, I spotted that, but figured it might be nice to make the local intention clearer, this should only be used by backups in normal, not a primary in normal (or we expect the hash chain to be valid).